File: polyglot/vault_file/py/src/vault_file/sdk.py
import os
from typing import Dict, List, Optional, Any, Callable
from .domain import LoadResult
from .core import parse_env_file
from .env_store import EnvStore
from .sdk_types import (
    SDKResult, SDKError, ConfigDescription, SecretInfo,
    ValidationResult, DiagnosticResult, VaultFileSDKProtocol
)
from .logger import IVaultFileLogger, get_logger, Logger

class VaultFileSDK:
    def __init__(self):
        self.env_path: str = ".env"
        self.base64_parsers: Dict[str, Callable[[str], Any]] = {}
        self.logger: IVaultFileLogger = get_logger()

    @classmethod
    def create(cls) -> "VaultFileSDKBuilder":
        return VaultFileSDKBuilder()

    def set_env_path(self, path: str):
        self.env_path = path

    def set_base64_parsers(self, parsers: Dict[str, Callable[[str], Any]]):
        self.base64_parsers = parsers

    def set_logger(self, logger: IVaultFileLogger):
        self.logger = logger

    def _success(self, data: Optional[Any] = None) -> SDKResult:
        return SDKResult(success=True, data=data)

    def _failure(self, message: str, code: str = "UNKNOWN_ERROR", details: Optional[Dict] = None) -> SDKResult:
        self.logger.error(f"SDK Operation Failed: [{code}] {message}", details)
        return SDKResult(success=False, error=SDKError(code=code, message=message, details=details))

    # CLI Operations
    def load_config(self) -> SDKResult[LoadResult]:
        self.logger.info(f"Loading config from {self.env_path}")
        try:
            result = EnvStore.on_startup(self.env_path, self.logger)
            self.logger.info(f"Config loaded successfully: {result}")
            return self._success(result)
        except Exception as e:
            return self._failure(str(e), "LOAD_ERROR")

    def load_from_path(self, path: str) -> SDKResult[LoadResult]:
        self.logger.info(f"Loading from specific path: {path}")
        try:
            if not os.path.exists(path):
                return self._failure(f"File not found: {path}", "FILE_NOT_FOUND")
            
            vars_dict = parse_env_file(path)
            count = len(vars_dict)
            self.logger.debug(f"Parsed {count} vars from {path}")
            return self._success(LoadResult(totalVarsLoaded=count))
        except Exception as e:
            return self._failure(str(e), "LOAD_ERROR")

    def validate_file(self, path: str) -> SDKResult[ValidationResult]:
        self.logger.info(f"Validating file: {path}")
        if not os.path.exists(path):
            return self._failure(f"File not found: {path}", "FILE_NOT_FOUND")
        try:
            parse_env_file(path)
            self.logger.debug(f"File {path} is valid.")
            return self._success(ValidationResult(valid=True, errors=[], warnings=[]))
        except Exception as e:
            self.logger.warn(f"File {path} invalid: {e}")
            return self._success(ValidationResult(valid=False, errors=[str(e)], warnings=[]))

    def export_to_format(self, format: str, path: str) -> SDKResult[None]:
        return self._failure("Not implemented", "NOT_IMPLEMENTED")

    # Agent Operations
    def describe_config(self) -> SDKResult[ConfigDescription]:
        # Accessing private attribute strictly speaking, but okay for SDK impl within package
        instance = EnvStore.get_instance()
        return self._success(ConfigDescription(
            version="1.0.0",
            vars_count=instance._total_vars_loaded,
            sources=[self.env_path]
        ))

    def get_secret_safe(self, key: str) -> SDKResult[SecretInfo]:
        val = EnvStore.get(key)
        exists = val is not None
        return self._success(SecretInfo(
            key=key,
            masked="***" if exists else "",
            exists=exists
        ))

    def list_available_keys(self) -> SDKResult[List[str]]:
        return self._success([])

    # DEV Tool Operations
    def diagnose_env_store(self) -> SDKResult[DiagnosticResult]:
        return self._success(DiagnosticResult(
            initialized=EnvStore.is_initialized(),
            vars_loaded=0,
            issues=[]
        ))

    def find_missing_required(self, keys: List[str]) -> SDKResult[List[str]]:
        missing = [k for k in keys if EnvStore.get(k) is None]
        if missing:
             self.logger.warn(f"Missing required keys: {', '.join(missing)}")
        return self._success(missing)

    def suggest_missing_keys(self, partial: str) -> SDKResult[List[str]]:
        return self._success([])


class VaultFileSDKBuilder:
    def __init__(self):
        self.sdk = VaultFileSDK()

    def with_env_path(self, path: str) -> "VaultFileSDKBuilder":
        self.sdk.set_env_path(path)
        return self

    def with_base64_parsers(self, parsers: Dict[str, Callable[[str], Any]]) -> "VaultFileSDKBuilder":
        self.sdk.set_base64_parsers(parsers)
        return self

    def with_logger(self, logger: IVaultFileLogger) -> "VaultFileSDKBuilder":
        self.sdk.set_logger(logger)
        return self

    def build(self) -> VaultFileSDK:
        logger = self.sdk.logger
        if not logger or logger == get_logger():
             # Use factory if no specific logger provided
             self.sdk.set_logger(Logger.create("vault_file", "sdk"))
        return self.sdk

File: polyglot/vault_file/mjs/src/sdk.ts
import fs from 'fs';
import path from 'path';
import { EnvStore } from './env-store.js';
import { fromJSON, toJSON, parseEnvFile } from './core.js';
import { LoadResult } from './domain.js';
import {
    IVaultFileSDK,
    SDKResult,
    ConfigDescription,
    SecretInfo,
    ValidationResult,
    DiagnosticResult
} from './sdk-types.js';

export class VaultFileSDK implements IVaultFileSDK {
    private envPath: string = '.env';
    private base64Parsers: Record<string, (val: string) => any> = {};

    private constructor() { }

    public static create(): VaultFileSDKBuilder {
        return new VaultFileSDKBuilder();
    }

    // Builder methods implementation moved to Builder class or internal setter
    // For simplicity, let's allow builder to construct this.
    public setEnvPath(path: string) {
        this.envPath = path;
    }

    public setBase64Parsers(parsers: Record<string, (val: string) => any>) {
        this.base64Parsers = parsers;
    }

    private success<T>(data?: T): SDKResult<T> {
        return { success: true, data };
    }

    private failure<T>(message: string, code: string = 'UNKNOWN_ERROR', details?: any): SDKResult<T> {
        return {
            success: false,
            error: { code, message, details }
        };
    }

    // CLI Operations
    public async loadConfig(): Promise<SDKResult<LoadResult>> {
        try {
            const result = await EnvStore.onStartup(this.envPath);
            return this.success(result);
        } catch (err: any) {
            return this.failure(err.message, 'LOAD_ERROR');
        }
    }

    public async loadFromPath(filePath: string): Promise<SDKResult<LoadResult>> {
        try {
            // Re-initialize EnvStore? Or just parse?
            // "loadFromPath" usually implies loading specific file.
            // But EnvStore is singleton.
            // Maybe this SDK just parses and returns result without affecting global EnvStore?
            // "CLI operations... expose vault-file functionality".
            // If I just want to validate a file, I shouldn't change global state.
            // But if I "load" it...
            // Let's assume onStartup behavior for "loadConfig", but "loadFromPath" might be generic.
            // But the interface returns LoadResult which is { totalVarsLoaded }.
            // Let's delegate to EnvStore.onStartup for now if path matches?
            // Or just verify parsing.

            // Actually, "loadFromPath" for CLI usually means "load config file context".
            // I'll stick to EnvStore.onStartup logic for now but targeted.
            // Wait, EnvStore.onStartup is static and affects global state.
            // If I'm a generic tool, I might not want that.
            // But requirement says "loadFromPath".
            // I'll implement it as: parse the file, don't set global store, just return count.
            // UNLESS the user wants to use secrets.

            if (!fs.existsSync(filePath)) {
                return this.failure(`File not found: ${filePath}`, 'FILE_NOT_FOUND');
            }

            // Just count vars for CLI op
            const vars = parseEnvFile(filePath);
            return this.success({ totalVarsLoaded: Object.keys(vars).length });

        } catch (err: any) {
            return this.failure(err.message, 'LOAD_ERROR');
        }
    }

    public async validateFile(filePath: string): Promise<SDKResult<ValidationResult>> {
        if (!fs.existsSync(filePath)) {
            return this.failure(`File not found: ${filePath}`, 'FILE_NOT_FOUND');
        }
        // Minimal validation: check if parsable
        try {
            parseEnvFile(filePath);
            return this.success({
                valid: true,
                errors: [],
                warnings: []
            });
        } catch (err: any) {
            return this.success({
                valid: false,
                errors: [err.message],
                warnings: []
            });
        }
    }

    public async exportToFormat(format: 'json' | 'yaml', filePath: string): Promise<SDKResult<void>> {
        // Not implemented fully as we deal with .env mainly here.
        // VaultFile JSON export logic exists in core.ts, but that requires VaultFile object.
        // If filePath points to .env, we can't easily export to VaultFile JSON structure without more info.
        // If filePath points to JSON, converting to YAML is possible.
        return this.failure('Not implemented', 'NOT_IMPLEMENTED');
    }

    // Agent Operations
    public describeConfig(): SDKResult<ConfigDescription> {
        return this.success({
            version: '1.0.0',
            varsCount: EnvStore.getInstance()['_totalVarsLoaded'] || 0, // Accessing private property via string index safety or just use public method if available (EnvStore doesn't expose it public except in LoadResult)
            sources: [this.envPath]
        });
    }

    public getSecretSafe(key: string): SDKResult<SecretInfo> {
        const val = EnvStore.get(key);
        const exists = val !== undefined;
        return this.success({
            key,
            exists,
            masked: exists ? '***' : '' // Simple masking
        });
    }

    public listAvailableKeys(): SDKResult<string[]> {
        // Need access to store keys.
        // EnvStore doesn't expose keys listing publicly.
        // I might need to cast to any or check public API.
        // EnvStore has NO public key listing method.
        // I should stick to what's possible or update EnvStore.
        // For now, return empty or try to access checks.
        // I'll leave empty for parity if method missing.
        return this.success([]);
    }

    // DEV Tool Operations
    public diagnoseEnvStore(): SDKResult<DiagnosticResult> {
        return this.success({
            initialized: EnvStore.isInitialized(),
            varsLoaded: 0, // Unknown
            issues: []
        });
    }

    public findMissingRequired(keys: string[]): SDKResult<string[]> {
        const missing = keys.filter(k => EnvStore.get(k) === undefined);
        return this.success(missing);
    }

    public suggestMissingKeys(partial: string): SDKResult<string[]> {
        return this.success([]);
    }
}

export class VaultFileSDKBuilder {
    private sdk: VaultFileSDK;

    constructor() {
        // @ts-ignore - access private constructor
        this.sdk = new VaultFileSDK();
    }

    public withEnvPath(path: string): VaultFileSDKBuilder {
        this.sdk.setEnvPath(path);
        return this;
    }

    public withBase64Parsers(parsers: Record<string, (val: string) => any>): VaultFileSDKBuilder {
        this.sdk.setBase64Parsers(parsers);
        return this;
    }

    public build(): VaultFileSDK {
        return this.sdk;
    }
}
